util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,458 @@
+		"/startHTTPServer": StartHTTPServer, "/startHTTPClient": StartHTTPClient, "/metrics": Metrics}
+	startListening(workerListenPort, h)
+	klog.Info("Started listening to Server")
+}
+
+func startListening(port string, handlers map[string]func(http.ResponseWriter, *http.Request)) {
+	klog.Info("In StartHTTPServer")
+	for p, h := range handlers {
+		http.HandleFunc(p, h)
+	}
+	err := http.ListenAndServe(":"+port, nil)
+	if err != nil {
+		klog.Fatalf("Failed starting http server for port: %v, Error: %v", port, err)
+	}
+}
+
+//Metrics returns the metrics collected
+func Metrics(res http.ResponseWriter, req *http.Request) {
+	var reply metricResponse
+	klog.Info("In metrics")
+	var stat string
+	select {
+	case stat = <-resultStatus:
+		if stat != "OK" {
+			klog.Error("Error collecting metrics:", stat)
+			reply.Error = "metrics collection failed:" + stat
+			createResp(reply, &res)
+			return
+		}
+		klog.Info("Metrics collected")
+	default:
+		klog.Info("Metric collection in progress")
+		reply.Error = "metric collection in progress"
+		createResp(reply, &res)
+		return
+	}
+	o, err := parseResult()
+	if err != nil {
+		reply.Error = err.Error()
+		createResp(reply, &res)
+		return
+	}
+	reply.Result = o
+	//startime and other variables below can be removed in future, for debugging during initial runs
+	stStr := strconv.FormatInt(startedAt, 10)
+	ftStr := strconv.FormatInt(futureTime, 10)
+	dtStr := strconv.FormatInt(futureTime-startedAt, 10)
+	reply.WorkerStartTime = "StartedAt:" + stStr + " FutureTime:" + ftStr + " diffTime:" + dtStr
+	createResp(reply, &res)
+}
+
+func createResp(resp interface{}, w *http.ResponseWriter) {
+	klog.Info("Inside reply")
+	(*w).Header().Set("Content-Type", "application/json")
+	(*w).WriteHeader(http.StatusOK)
+	klog.Info("Marshalled Resp:", resp)
+	b, err := json.Marshal(resp)
+	if err != nil {
+		klog.Info("Error marshalling to json:", err)
+	}
+	(*w).Write(b)
+}
+
+//StartTCPClient starts iperf client for tcp measurements
+func StartTCPClient(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartTCPClient")
+	klog.Info("Req:", req)
+	ts, dur, destIP, _ := parseURLParam(req)
+	if dur == "" || destIP == "" {
+		createResp(startWrkResponse{error: "missing/invalid required parameters"}, &res)
+		return
+	}
+	go schedule(protocolTCP, ts, dur,
+		"iperf", []string{"-c", destIP, "-f", "K", "-l",
+			"20", "-b", "1M", "-i", "1", "-t", dur})
+	createResp(startWrkResponse{}, &res)
+}
+
+//StartTCPServer starts iperf server for tcp measurements
+func StartTCPServer(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartTCPServer")
+	klog.Info("Req:", req)
+	ts, dur, _, numcl := parseURLParam(req)
+	if dur == "" || numcl == "" {
+		createResp(startWrkResponse{error: "missing/invalid required parameters"}, &res)
+		return
+	}
+	go schedule(protocolTCP, ts, dur, "iperf", []string{"-s", "-f", "K", "-i", dur, "-P", numcl})
```

Comment:
also, this iperf -s will never die, I didn't go further in the code, sorry, how do you exit it?

Reply:
We are giving option to wait for specified num of clients before closing, hence after the required number of clients have connected(and collected metrics), iperf would close the server.